source/windows/secure_channel_tls_handler.c

```diff
@@ -1585,11 +1558,17 @@ static int s_handler_shutdown(
                 struct aws_io_message *outgoing_message = aws_channel_acquire_message_from_pool(
                     slot->channel, AWS_IO_MESSAGE_APPLICATION_DATA, output_buffer.cbBuffer);
 
-                if (!outgoing_message || outgoing_message->message_data.capacity < output_buffer.cbBuffer) {
-                    return aws_channel_slot_on_handler_shutdown_complete(slot, dir, aws_last_error(), true);
+                if (outgoing_message->message_data.capacity < output_buffer.cbBuffer) {
+                    aws_mem_release(outgoing_message->allocator, outgoing_message);
+                    FreeContextBuffer(output_buffer.pvBuffer);
+                    if (error_code == 0) {
+                        error_code = AWS_IO_TLS_ERROR_WRITE_FAILURE;
+                    }
+                    return aws_channel_slot_on_handler_shutdown_complete(slot, dir, error_code, true);
                 }
                 memcpy(outgoing_message->message_data.buffer, output_buffer.pvBuffer, output_buffer.cbBuffer);
                 outgoing_message->message_data.len = output_buffer.cbBuffer;
+                FreeContextBuffer(output_buffer.pvBuffer);

```

Comment:
why we only free the buffer when `status == SEC_E_OK || status == SEC_I_CONTEXT_EXPIRED`?
The [doc](https://learn.microsoft.com/en-us/windows/win32/api/sspi/nf-sspi-initializesecuritycontexta) doesn't seem to be very clear about failure, but it does mention that it may contain alert information with failure, if i understand correctly
```
On output, if an alert is generated, this buffer contains information about that alert, and the function fails.
```

Reply:
Should be a typo by @graebm . In my original commit #651 , FreeContextBuffer is called before return of function, no matter success or failure.

Reply:
Thanks for pointing that out, both of you.

The existing code had several functions where `FreeContextBuffer()` wasn't being called in call cases after a [AcceptSecurityContext()](https://learn.microsoft.com/en-us/windows/win32/api/sspi/nf-sspi-acceptsecuritycontext) or [InitializeSecurityContextA()](https://learn.microsoft.com/en-us/windows/win32/api/sspi/nf-sspi-initializesecuritycontexta) call.

I changed the error-handling styles for these functions, so `FreeContextBuffer()` is **always** called, regardless of what gets returned from `AcceptSecurityContext()`/`InitializeSecurityContextA()`.